core/os/os2/process.odin

```diff
@@ -392,44 +394,54 @@ process_exec :: proc(
 		desc.stderr = stderr_w
 		process = process_start(desc) or_return
 	}
+
 	stdout_builder := strings.builder_make(allocator) or_return
 	stderr_builder := strings.builder_make(allocator) or_return
-	read_data: for {
-		buf: [1024]u8
+
+	has_stdout, has_stderr: bool
+	read_data: for !has_stdout || !has_stderr {
+		buf: [1024]u8 = ---
 		n: int
 		has_data: bool
-		hangup := false
-		has_data, err = pipe_has_data(stdout_r)
-		if has_data {
-			n, err = read(stdout_r, buf[:])
-			strings.write_bytes(&stdout_builder, buf[:n])
-		}
-		switch err {
-		case nil: // nothing
-		case .Broken_Pipe:
-			hangup = true
-		case:
-			return
-		}
-		has_data, err = pipe_has_data(stderr_r)
-		if has_data {
-			n, err = read(stderr_r, buf[:])
-			strings.write_bytes(&stderr_builder, buf[:n])
-		}
-		switch err {
-		case nil: // nothing
-		case .Broken_Pipe:
-			hangup = true
-		case:
-			return
+
+		if !has_stdout {
+			has_data, err = pipe_has_data(stdout_r)
+			if has_data {
+				n, err = read(stdout_r, buf[:])
+				if strings.write_bytes(&stdout_builder, buf[:n]) != n {
+					err = .Out_Of_Memory
```

Comment:
This shouldn't return out of memory, because at the end we're pushing data to a string builder. Is there a way we can limit how much bytes we're reading or am I misreading what that does?

Reply:
It is returning out of memory when the amount of bytes written to the string builder is not the same as the amount we try to write. I think its safe to assume you are out of memory then (the string builder didn't grow correctly).

Reply:
I realised I probably misread this code. Alright then

Reply:
I changed it to just a dynamic array instead of a string builder to be clearer